File: mobile/lib/services/review_service.dart
import 'package:riverpod_annotation/riverpod_annotation.dart';
import '../models/review.dart';
import '../providers/auth_provider.dart';
import 'api_client.dart';

part 'review_service.g.dart';

@riverpod
ReviewService reviewService(ReviewServiceRef ref) {
  final apiClient = ref.watch(apiClientProvider);
  return ReviewService(apiClient);
}

class ReviewService {
  final ApiClient _apiClient;

  ReviewService(this._apiClient);

  // Create a review
  Future<Review> createReview(CreateReviewData data) async {
    final response = await _apiClient.post(
      '/reviews',
      data: {
        'requestId': data.requestId,
        'score': data.score,
        if (data.comment != null) 'comment': data.comment,
        if (data.qualities != null)
          'qualities': {
            if (data.qualities!.punctuality != null)
              'punctuality': data.qualities!.punctuality,
            if (data.qualities!.professionalism != null)
              'professionalism': data.qualities!.professionalism,
            if (data.qualities!.quality != null)
              'quality': data.qualities!.quality,
            if (data.qualities!.cleanliness != null)
              'cleanliness': data.qualities!.cleanliness,
            if (data.qualities!.communication != null)
              'communication': data.qualities!.communication,
          },
        'images': data.images,
      },
    );
    return Review.fromJson(response.data['data']);
  }

  // Get review by ID
  Future<Review> getReviewById(String reviewId) async {
    final response = await _apiClient.get('/reviews/$reviewId');
    return Review.fromJson(response.data['data']);
  }

  // Get craftsman reviews
  Future<ReviewListResponse> getCraftsmanReviews(
    String craftsmanId, {
    int page = 1,
    int limit = 20,
    int? minScore,
    String sortBy = 'createdAt',
    String sortOrder = 'desc',
  }) async {
    final params = <String, dynamic>{
      'page': page,
      'limit': limit,
      'sortBy': sortBy,
      'sortOrder': sortOrder,
    };
    if (minScore != null) {
      params['minScore'] = minScore;
    }

    final response = await _apiClient.get(
      '/reviews/craftsman/$craftsmanId',
      queryParameters: params,
    );

    return ReviewListResponse(
      data: (response.data['data'] as List)
          .map((e) => Review.fromJson(e))
          .toList(),
      page: response.data['pagination']['page'],
      totalPages: response.data['pagination']['totalPages'],
      total: response.data['pagination']['total'],
      limit: response.data['pagination']['limit'],
    );
  }

  // Get craftsman reviews summary
  Future<ReviewsSummary> getCraftsmanReviewsSummary(String craftsmanId) async {
    final response =
        await _apiClient.get('/reviews/craftsman/$craftsmanId/summary');
    return ReviewsSummary.fromJson(response.data['data']);
  }

  // Get my reviews
  Future<ReviewListResponse> getMyReviews({
    int page = 1,
    int limit = 20,
  }) async {
    final response = await _apiClient.get(
      '/reviews/my',
      queryParameters: {'page': page, 'limit': limit},
    );

    return ReviewListResponse(
      data: (response.data['data'] as List)
          .map((e) => Review.fromJson(e))
          .toList(),
      page: response.data['pagination']['page'],
      totalPages: response.data['pagination']['totalPages'],
      total: response.data['pagination']['total'],
      limit: response.data['pagination']['limit'],
    );
  }

  // Update review
  Future<Review> updateReview(
    String reviewId, {
    String? comment,
    List<String>? images,
  }) async {
    final response = await _apiClient.put(
      '/reviews/$reviewId',
      data: {
        if (comment != null) 'comment': comment,
        if (images != null) 'images': images,
      },
    );
    return Review.fromJson(response.data['data']);
  }

  // Respond to review (craftsman)
  Future<Review> respondToReview(String reviewId, String response) async {
    final res = await _apiClient.post(
      '/reviews/$reviewId/respond',
      data: {'response': response},
    );
    return Review.fromJson(res.data['data']);
  }

  // Report a review
  Future<void> reportReview(String reviewId, String reason) async {
    await _apiClient.post(
      '/reviews/$reviewId/report',
      data: {'reason': reason},
    );
  }
}

File: backend/src/services/review.service.ts
import { Types } from 'mongoose';
import Review, { IReview } from '@models/Review';
import ServiceRequest from '@models/ServiceRequest';
import Craftsman from '@models/Craftsman';
import Customer from '@models/Customer';
import { NotFoundError, BadRequestError, ForbiddenError } from '@utils/errors';

interface CreateReviewParams {
  requestId: string;
  customerId: string;
  score: number;
  comment?: string;
  qualities?: {
    punctuality?: number;
    professionalism?: number;
    quality?: number;
    cleanliness?: number;
    communication?: number;
  };
  images?: string[];
}

interface GetReviewsParams {
  craftsmanId?: string;
  customerId?: string;
  minScore?: number;
  maxScore?: number;
  page?: number;
  limit?: number;
  sortBy?: 'createdAt' | 'score';
  sortOrder?: 'asc' | 'desc';
}

class ReviewService {
  /**
   * Create a review for a completed request
   */
  async createReview(params: CreateReviewParams): Promise<IReview> {
    const { requestId, customerId, score, comment, qualities, images } = params;

    // Validate request exists and is completed
    const request = await ServiceRequest.findById(requestId);
    if (!request) {
      throw new NotFoundError('الطلب غير موجود');
    }

    if (request.status !== 'completed') {
      throw new BadRequestError('لا يمكن تقييم طلب غير مكتمل');
    }

    // Validate customer owns the request
    const customer = await Customer.findById(customerId);
    if (!customer || request.customerId.toString() !== customerId) {
      throw new ForbiddenError('ليس لديك صلاحية لتقييم هذا الطلب');
    }

    // Check if already reviewed
    const existingReview = await Review.findOne({
      requestId: new Types.ObjectId(requestId),
    });
    if (existingReview) {
      throw new BadRequestError('تم تقييم هذا الطلب مسبقاً');
    }

    // Get craftsman ID
    if (!request.craftsmanId) {
      throw new BadRequestError('لا يوجد صنايعي مخصص لهذا الطلب');
    }

    const craftsmanId = request.craftsmanId;

    // Create review
    const review = await Review.create({
      requestId: new Types.ObjectId(requestId),
      customerId: new Types.ObjectId(customerId),
      craftsmanId: new Types.ObjectId(craftsmanId),
      score,
      comment,
      qualities: qualities || {},
      images: images || [],
    });

    // Update craftsman's rating
    await this.updateCraftsmanRating(craftsmanId.toString());

    // Populate and return
    return this.getReviewById(review._id.toString());
  }

  /**
   * Get review by ID
   */
  async getReviewById(reviewId: string): Promise<IReview> {
    const review = await Review.findById(reviewId)
      .populate({
        path: 'customerId',
        select: 'userId',
        populate: {
          path: 'userId',
          select: 'name avatar',
        },
      })
      .populate({
        path: 'requestId',
        select: 'requestNumber title categoryId',
        populate: {
          path: 'categoryId',
          select: 'nameAr',
        },
      });

    if (!review) {
      throw new NotFoundError('التقييم غير موجود');
    }

    return review;
  }

  /**
   * Get reviews with filters
   */
  async getReviews(params: GetReviewsParams) {
    const {
      craftsmanId,
      customerId,
      minScore,
      maxScore,
      page = 1,
      limit = 20,
      sortBy = 'createdAt',
      sortOrder = 'desc',
    } = params;

    const query: Record<string, unknown> = {
      isVisible: true,
    };

    if (craftsmanId) {
      query.craftsmanId = new Types.ObjectId(craftsmanId);
    }

    if (customerId) {
      query.customerId = new Types.ObjectId(customerId);
    }

    if (minScore !== undefined || maxScore !== undefined) {
      query.score = {};
      if (minScore !== undefined) {
        (query.score as Record<string, number>).$gte = minScore;
      }
      if (maxScore !== undefined) {
        (query.score as Record<string, number>).$lte = maxScore;
      }
    }

    const skip = (page - 1) * limit;
    const sort: Record<string, 1 | -1> = {
      [sortBy]: sortOrder === 'asc' ? 1 : -1,
    };

    const [reviews, total] = await Promise.all([
      Review.find(query)
        .sort(sort)
        .skip(skip)
        .limit(limit)
        .populate({
          path: 'customerId',
          select: 'userId',
          populate: {
            path: 'userId',
            select: 'name avatar',
          },
        })
        .populate({
          path: 'requestId',
          select: 'requestNumber title categoryId',
          populate: {
            path: 'categoryId',
            select: 'nameAr',
          },
        }),
      Review.countDocuments(query),
    ]);

    return {
      data: reviews,
      pagination: {
        page,
        limit,
        total,
        totalPages: Math.ceil(total / limit),
      },
    };
  }

  /**
   * Get craftsman reviews summary
   */
  async getCraftsmanReviewsSummary(craftsmanId: string) {
    const reviews = await Review.find({
      craftsmanId: new Types.ObjectId(craftsmanId),
      isVisible: true,
    });

    if (reviews.length === 0) {
      return {
        averageRating: 0,
        totalReviews: 0,
        ratingDistribution: { 1: 0, 2: 0, 3: 0, 4: 0, 5: 0 },
        qualityAverages: {
          punctuality: 0,
          professionalism: 0,
          quality: 0,
          cleanliness: 0,
          communication: 0,
        },
      };
    }

    // Calculate average rating
    const totalScore = reviews.reduce((sum, r) => sum + r.score, 0);
    const averageRating = totalScore / reviews.length;

    // Calculate rating distribution
    const ratingDistribution = { 1: 0, 2: 0, 3: 0, 4: 0, 5: 0 };
    reviews.forEach((r) => {
      ratingDistribution[r.score as keyof typeof ratingDistribution]++;
    });

    // Calculate quality averages
    const qualitySums = {
      punctuality: 0,
      professionalism: 0,
      quality: 0,
      cleanliness: 0,
      communication: 0,
    };
    const qualityCounts = { ...qualitySums };

    reviews.forEach((r) => {
      if (r.qualities) {
        Object.keys(qualitySums).forEach((key) => {
          const k = key as keyof typeof qualitySums;
          const qualityValue = r.qualities[k];
          if (qualityValue !== undefined && qualityValue !== null) {
            qualitySums[k] += qualityValue;
            qualityCounts[k]++;
          }
        });
      }
    });

    const qualityAverages = Object.keys(qualitySums).reduce(
      (acc, key) => {
        const k = key as keyof typeof qualitySums;
        acc[k] = qualityCounts[k] > 0 ? qualitySums[k] / qualityCounts[k] : 0;
        return acc;
      },
      {} as Record<string, number>
    );

    return {
      averageRating: Math.round(averageRating * 10) / 10,
      totalReviews: reviews.length,
      ratingDistribution,
      qualityAverages,
    };
  }

  /**
   * Update review (only comment/images within time limit)
   */
  async updateReview(
    reviewId: string,
    customerId: string,
    updates: { comment?: string; images?: string[] }
  ): Promise<IReview> {
    const review = await Review.findOne({
      _id: new Types.ObjectId(reviewId),
      customerId: new Types.ObjectId(customerId),
    });

    if (!review) {
      throw new NotFoundError('التقييم غير موجود أو ليس لديك صلاحية التعديل');
    }

    // Check if within 24 hours
    const hoursSinceCreation =
      (Date.now() - review.createdAt.getTime()) / (1000 * 60 * 60);
    if (hoursSinceCreation > 24) {
      throw new BadRequestError('لا يمكن تعديل التقييم بعد 24 ساعة');
    }

    if (updates.comment !== undefined) {
      review.comment = updates.comment;
    }

    if (updates.images !== undefined) {
      review.images = updates.images;
    }

    await review.save();

    return this.getReviewById(reviewId);
  }

  /**
   * Craftsman response to a review
   */
  async respondToReview(
    reviewId: string,
    craftsmanId: string,
    response: string
  ): Promise<IReview> {
    const review = await Review.findOne({
      _id: new Types.ObjectId(reviewId),
      craftsmanId: new Types.ObjectId(craftsmanId),
    });

    if (!review) {
      throw new NotFoundError('التقييم غير موجود أو ليس لديك صلاحية الرد');
    }

    if (review.craftsmanResponse) {
      throw new BadRequestError('تم الرد على هذا التقييم مسبقاً');
    }

    review.craftsmanResponse = response;
    review.respondedAt = new Date();
    await review.save();

    return this.getReviewById(reviewId);
  }

  /**
   * Report a review
   */
  async reportReview(
    reviewId: string,
    _reporterId: string,
    reason: string
  ): Promise<void> {
    const review = await Review.findById(reviewId);

    if (!review) {
      throw new NotFoundError('التقييم غير موجود');
    }

    review.isReported = true;
    review.reportReason = reason;
    await review.save();

    // TODO: Notify admin for review
  }

  /**
   * Update craftsman's overall rating
   */
  private async updateCraftsmanRating(craftsmanId: string): Promise<void> {
    const result = await Review.aggregate([
      {
        $match: {
          craftsmanId: new Types.ObjectId(craftsmanId),
          isVisible: true,
        },
      },
      {
        $group: {
          _id: null,
          averageRating: { $avg: '$score' },
          totalRatings: { $sum: 1 },
        },
      },
    ]);

    if (result.length > 0) {
      await Craftsman.findByIdAndUpdate(craftsmanId, {
        rating: Math.round(result[0].averageRating * 10) / 10,
        totalRatings: result[0].totalRatings,
      });
    }
  }

  /**
   * Admin: Hide/Show review
   */
  async toggleReviewVisibility(
    reviewId: string,
    isVisible: boolean
  ): Promise<IReview> {
    const review = await Review.findByIdAndUpdate(
      reviewId,
      { isVisible },
      { new: true }
    );

    if (!review) {
      throw new NotFoundError('التقييم غير موجود');
    }

    // Update craftsman rating
    await this.updateCraftsmanRating(review.craftsmanId.toString());

    return this.getReviewById(reviewId);
  }
}

export const reviewService = new ReviewService();
